From ba7f8b00168a288b8d64404feb154494c77d035e Mon Sep 17 00:00:00 2001
From: Bogdan-Gabriel Roman <bogdan-gabriel.roman@nxp.com>
Date: Thu, 13 May 2021 00:07:40 +0300
Subject: [PATCH 233/269] bl2_el3: resume bl31 earlier when waking up from
 standby

There is no need for setting up io devices in BL2 (s32g_io_setup()) when
coming out of standby. Start resuming bl31 before attempting this.
Moreover, doing s32g_io_setup() when coming out of standby leads to
nefarious consequences: the boot source is not correctly identified
because BOOT_GPR_BMR1 register is not re-popultaed with the RCON pads
values upon coming out of standby. Thus, s32g_io_setup() would crash.

Issue: ALB-7120
Upstream-Status: Pending 

Signed-off-by: Bogdan-Gabriel Roman <bogdan-gabriel.roman@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 plat/nxp/s32g/s32g_bl2_el3.c | 43 ++++++++++++++++++------------------
 1 file changed, 22 insertions(+), 21 deletions(-)

diff --git a/plat/nxp/s32g/s32g_bl2_el3.c b/plat/nxp/s32g/s32g_bl2_el3.c
index 31209c72c..75733c7af 100644
--- a/plat/nxp/s32g/s32g_bl2_el3.c
+++ b/plat/nxp/s32g/s32g_bl2_el3.c
@@ -194,25 +194,6 @@ struct bl_load_info *plat_get_bl_image_load_info(void)
 	return get_bl_load_info_from_mem_params_desc();
 }
 
-void bl2_el3_early_platform_setup(u_register_t arg0, u_register_t arg1,
-				  u_register_t arg2, u_register_t arg3)
-{
-	size_t index;
-	bl_mem_params_node_t *params = s32g_bl2_mem_params_descs;
-
-	s32g_early_plat_init(false);
-	s32g_io_setup();
-
-	add_fip_img_to_mem_params_descs(params, &index);
-	add_bl31_img_to_mem_params_descs(params, &index);
-	add_bl32_img_to_mem_params_descs(params, &index);
-	add_bl32_extra1_img_to_mem_params_descs(params, &index);
-	add_bl33_img_to_mem_params_descs(params, &index);
-	add_invalid_img_to_mem_params_descs(params, &index);
-
-	bl_mem_params_desc_num = index;
-}
-
 static int disable_clk_node(void *blob, uint32_t *phandle)
 {
 	const char *clk_path;
@@ -576,10 +557,14 @@ void s32g_el3_mmu_fixup(void)
 	enable_mmu_el3(0);
 }
 
-void bl2_el3_plat_arch_setup(void)
+void bl2_el3_early_platform_setup(u_register_t arg0, u_register_t arg1,
+				  u_register_t arg2, u_register_t arg3)
 {
+	size_t index;
+	bl_mem_params_node_t *params = s32g_bl2_mem_params_descs;
 	struct s32g_ssram_mailbox *ssram_mb = (void *)BL31SSRAM_MAILBOX;
-	uint32_t ret;
+
+	s32g_early_plat_init(false);
 
 	console_s32g_register();
 
@@ -590,6 +575,22 @@ void bl2_el3_plat_arch_setup(void)
 		panic();
 	}
 
+	s32g_io_setup();
+
+	add_fip_img_to_mem_params_descs(params, &index);
+	add_bl31_img_to_mem_params_descs(params, &index);
+	add_bl32_img_to_mem_params_descs(params, &index);
+	add_bl32_extra1_img_to_mem_params_descs(params, &index);
+	add_bl33_img_to_mem_params_descs(params, &index);
+	add_invalid_img_to_mem_params_descs(params, &index);
+
+	bl_mem_params_desc_num = index;
+}
+
+void bl2_el3_plat_arch_setup(void)
+{
+	uint32_t ret;
+
 	s32g_el3_mmu_fixup();
 
 	s32g_sram_clear(S32G_BL33_IMAGE_BASE, DTB_BASE);
-- 
2.17.1

